### ISSUE TITLE
Write Support

### ISSUE THREAD

**Issue by ibotty** [duckdb/duckdb-iceberg#37]
Is there write support on the roadmap?  If so, is there an ETA?  I'll have to plan a future project and would like to keep using duckdb.

**Comment by samansmink** [comment]
Hi @ibotty! It's something we want to add but we can not give an ETA currently

**Comment by ibotty** [comment]
Thank you for the answer.  I might consider duckdb when time comes.

**Comment by aaronsteers** [comment]
Hi, @samansmink . I'm interested also - and I might have cycles to contribute. Is it correct that PyIceberg is not a blocker here? I think I recall we previously were waiting for write support from PyIceberg - but it looks like they support writing now. Sure that sound right?

Thanks!

**Comment by rustyconover** [comment]
Incrementally, this is getting closer with the COPY function in DuckDB adding support for writing field id metadata for the columns.

There should be an `iceberg_schema()` function that will return the schema of a table and the appropriate field ids.  Then you could write a Parquet file with the right field ids to be added to the Iceberg table.

**Comment by NielsKorschinsky** [comment]
Hi @rustyconover ,

that are great news!
Do I understand you correctly, that this `iceberg_schema()` function is already available (in a pre-release)?
What is the plan with writing the associated parquet file, could this also be done in sync with writing the schema file via duckdb?

**Comment by rustyconover** [comment]
Hi @NielsKorschinsky,

At this current time the function `iceberg_schema()` doesn't exist in the extension, but I can see its usefulness to be along the road of writing changes to Iceberg tables.

Rusty

**Comment by NicolasPA** [comment]
I'd like to test DuckDB on Iceberg for my ELT SQL processing engine on cloud storage for our DWH migration, and I bet many more people do. So here's a more precise use case to motivate the prioritization of this feature.

**Comment by NicolasPA** [comment]
Hi @samansmink, I believe this request is the most voted of all the DuckDB repos.  
Would you have an updated ETA motivated by this demand?
We want to know if we can count on DuckDB for our new data platform.

**Comment by samansmink** [comment]
@NicolasPA I can assure you that the interest for this feature has not gone unnoticed. However, at this time we do not have an ETA for this feature yet.

I hope you understand that as a relatively small [non vc-backed](https://www.theregister.com/2023/10/05/duckdb_labs_puts_limit_on_vc_funds/) company, we don't always have the resources to build everything we want right away. Especially in the case of a feature with the complexity of this issue. 

For more info on our support policy, check out https://duckdblabs.com/news/2023/10/02/support-policy.html

**Comment by NicolasPA** [comment]
@samansmink, thank you for confirming you have been noticing the interest.
I fully understand the resources issue.  
I don't want to put unnecessary pressure on you guys.  
I think I was voicing the data architectures wet dreams I'm having with DuckDB lately, and hoping this kind of testimony will help you prioritize in the future.

**Comment by chlimaferreira** [comment]
I'm sure if duckdb writes in iceberg or delta it will become a big player in the market.

**Comment by ani-panda** [comment]
This issue is open for a very long time. I wonder if Tabular (https://www.tabular.io/) folks are doing anything? That said, Databricks acquisition of Tabular may not be motivating Databricks to cannibalize their own (spark) engine.

**Comment by pantonis** [comment]
+1

**Comment by thenaturalist** [comment]
+1

**Comment by 2tony2** [comment]
This is the biggest factor holding back using duckdb as an actual data warehouse replacement as far as I'm concerned.

**Comment by beeing** [comment]
I noticed that there's https://github.com/apache/iceberg-rust which is under active development, perhaps it will be easier to wait for this library and integrate to this extension.

**Comment by sergun** [comment]
Vote for this feature

**Comment by guizmaii** [comment]
+1 🙂

**Comment by samansmink** [comment]
Hey everyone! thanks for participating in the discussion and letting us know the importance of this feature to you. I can assure you this has not gone unnoticed and we are very busy with getting this on the roadmap. 

Please be considerate of the fact that every `+1` comment will send a notification to all 17 people subscribed to this issue and consider replying to the initial issue with a 👍. We regularly scan our open issues sorted by upvotes to get a an understanding of what the community wants

**Comment by gtonietti-demyst** [comment]
The addition of this feature would be world changing, the cherry on top of an already excellent piece of technology.

Here you have my vote 👍

**Comment by g-m-bell** [comment]
This is informing critical engineering decisions at my company. +1

**Comment by stenvala** [comment]
Started checking if I could create a simple demo crud app with duckdb and s3 table bucket ... But I didn't get the data ever in. Perhaps not the tech stack to investigate more. Great work though!

**Comment by abrunner94** [comment]
Check out Delta Lake + Datafusion

**Comment by abrunner94** [comment]
You don't actually need S3 table buckets for it, regular S3 bucket is sufficient

Delta Lake has functionality for vacuuming and compacting files, S3 Table buckets charge for this if I'm not mistaken

**Comment by stenvala** [comment]
True, but I especially wanted to build on table buckets just to study how stuff would work with them

**Comment by jovezhong** [comment]
Given the fact that there is no ready-to-use open-source C++ SDK for Iceberg write, almost all C++ based OLAP (e.g. DuckDB, ClickHouse) are focusing on reading data. But as you see those passionate comments above, still many ppl want a clean/easy way to write to Iceberg tables, ideally as easy as using DuckDB, not SparkSQL or AWS binding. My impression is the https://github.com/apache/iceberg-cpp is still in early stage, and there is a cool idea to use Rust to implement this then provide a C binding (https://github.com/apache/iceberg-rust/pull/966 ) 

Spoiler alert: later this month (hopefully next week), my team will open-source our C++ code to write to Iceberg tables, with initial support of REST catalog only, under our OSS repo https://github.com/timeplus-io/proton In the end, you will just need to drop a binary file somewhere and use SQL to connect to Iceberg/catalog and read/write data.

**Comment by mike-luabase** [comment]
@jovezhong did this happen? I don't see much about Iceberg in the repo you linked to.

**Comment by jovezhong** [comment]
Hi mike, here is the PR to add Iceberg external stream to Timeplus Proton https://github.com/timeplus-io/proton/pull/928 It's not merged yet but feel free to review the code and share your thoughts

**Comment by Tishj** [comment]
Work has started on adding write support, the first PR, which adds support for the INSERT statement (unpartitioned only for now) targeting Iceberg tables, is here: https://github.com/duckdb/duckdb-iceberg/pull/305

**Comment by szarnyasg** [comment]
FYI, DuckDB also has a roadmap now, see: https://duckdb.org/roadmap

**Comment by Tishj** [comment]
This has since been implemented, we will even have INSERT+UPDATE support for V3 tables starting in v1.5.1 👍
